numpy/core/src/multiarray/datetime.c

```diff
@@ -1761,6 +1762,9 @@ parse_datetime_unit_from_string(char *str, Py_ssize_t len, char *metastr)
                 return NPY_FR_as;
         }
     }
+    else if (allow_generic && !strcmp(str, "generic")) {
```

Comment:
You need to check `len` here, the string is not necessarily zero terminated, so maybe something like
```
else if (allow_generic && len == 7 && !strncmp(str, "generic", 7))
```